---
title: "Analysis of Orbitrap Fusion Data from Paulo *et al.*"
author: "Christopher Hughes"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

###Description

This file contains the code for processing the search and quantification output files from the yeast carbon source publication from Paulo *et al.* (PMID: 26399295). **These data were used to make Figure S-7.**

  * MGF files were generated using RawQuant.
  * The data were searched using SearchCLI and PeptideShakerCLI.
  * Raw files were quantified with RawQuant.
  
###Analysis

First call the required libraries.

```{r, message=FALSE}
library(Biostrings)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(cowplot)
library(reshape)
library(data.table)
library(gplots)
```

Next we want to make an index based on our search database that we can use to add descriptive data to our search result.

```{r}
####
yeastDB = readAAStringSet('/projects/ptx_analysis/chughes/databases/uniprot_s288c-crap-oct2017-FWD.fasta')
#make an annotation index from yeastDB
yeastIndex = data.frame('meta' = names(yeastDB))
yeastIndex$accession = sub(".*?\\|(.*?)\\|.*", "\\1", yeastIndex$meta) #grab the accession from the header
yeastIndex$gene = sub(".*?GN=(.*?) .*", "\\1", yeastIndex$meta) #grab the gene from the header
yeastIndex$species = sub(".*?OS=(.*?) [GP].*", "\\1", yeastIndex$meta) #grab the species from the header
yeastIndex = yeastIndex[,2:4]
```

Now we can go ahead and grab the identification and quantification data.

```{r}
#find the files of interest
#psms files from SearchCLI and PeptideShaker
psm_in = read.table('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/reports-output/paulo_yeast-carbon_all-fractions_1_Default_PSM_Report.txt', sep='\t', header=TRUE)
#quant files from RawQuant algorithm
quant_files = list.files('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/quant-output',pattern='*QuantData.txt',full.names=TRUE)
ms1_quant_files = list.files('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/quant-output',pattern='*MS1ParseData.txt',full.names=TRUE)
ms2_quant_files = list.files('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/quant-output',pattern='*MS2ParseData.txt',full.names=TRUE)
```

Now we can process the peptide spectral match data to a more manageable data frame, and incorporate the quantification data.

```{r, message=FALSE}
###this first section handles the identification data
psm_f1 = psm_in %>% 
  dplyr::select(Protein.s., Sequence, Spectrum.File, Spectrum.Title, m.z, Identification.Charge) %>% #select columns
  dplyr::mutate(Spectrum.Title, scan_num = sub(".*?spectrum (.*?)", "\\1", Spectrum.Title)) %>% #parse out the scan number
  dplyr::mutate(Identification.Charge, id_charge = gsub("[^0-9]", "", Identification.Charge)) %>% #parse out the charge
  dplyr::mutate(Spectrum.Title, fraction = sub("(.*?)_MGF\\.mgf", "\\1", Spectrum.File)) %>% #parse the fraction
  dplyr::select(Protein.s., Sequence, scan_num, m.z, id_charge, fraction) %>% #select new columns
  dplyr::mutate(Protein.s., first_accession = sapply(strsplit(as.character(Protein.s.),"\\,"), `[`, 1)) #parse the accession
#adjust the column names
colnames(psm_f1) = c('accession_groups','sequence','MS2ScanNumber','id_precursor_mz','id_charge','fraction','accession')
psm_f2 = merge(psm_f1,yeastIndex,by='accession',all.x=TRUE) #add the descriptive data
psm_f3 = dplyr::filter(psm_f2, !grepl('CONT', accession)) #remove contaminant hits
  
  
####this second section handles the RawQuant quantification data
all_quant_files = data.frame() #create an empty data frame to hold the output data
for (i in 1:length(quant_files)){
  quant_in = read.table(quant_files[i],header=TRUE,sep='\t') #grab the input file
  quant_in[,11:90][quant_in[,11:90] == 0] = NA #transform any zero or missing values to NA
  quant_meta = quant_in[,c('PrecursorMass','ScanNumber','MS1ScanNumber','MS2ScanNumber','RetentionTime',
                           'PrecursorCharge','MS3IonInjectionTime','MS1Interference','tmt126_mass','tmt127N_mass',
                           'tmt127C_mass','tmt128N_mass','tmt128C_mass','tmt129N_mass','tmt129C_mass',
                           'tmt130N_mass','tmt130C_mass','tmt131_mass')] #extract the quantification meta data
  quant_exprs = quant_in[,c('tmt126_intensity','tmt127N_intensity','tmt127C_intensity','tmt128N_intensity',
                            'tmt128C_intensity','tmt129N_intensity','tmt129C_intensity','tmt130N_intensity',
                            'tmt130C_intensity','tmt131_intensity','tmt126_noise','tmt127N_noise',
                            'tmt127C_noise','tmt128N_noise','tmt128C_noise','tmt129N_noise','tmt129C_noise','tmt130N_noise',
                            'tmt130C_noise','tmt131_noise')] #extract the quantification data
  quant_sn = quant_exprs[,c(1:10)] / quant_exprs[,c(11:20)] #calculate the signal to noise
  quant_sn$totalIntensity = rowSums(quant_exprs[,1:10],na.rm=TRUE) #calculate the total sum intensity
  quant_sn$totalSN = rowSums(quant_sn[,1:10],na.rm=TRUE) #calculate the total sum signal to noise
  quant_sn$totalNoise = rowSums(quant_exprs[,11:20],na.rm=TRUE) #calculate the total sum noise
  quant_sn$totalSPS = rowSums(quant_in[,81:90],na.rm=TRUE) #calculate the total sum SPS ion intensity
  quant_sn$SPScount = rowSums(!is.na(quant_in[,71:80])) #calculate the number of SPS ions selected
  quant_sn$totalNA = rowSums(is.na(quant_sn[,1:10])) #grab the number of NA values across the TMT channels
  quant_meta$fraction = sub(".*?quant-output\\/(.*?)_Quant.*", "\\1", quant_files[i]) #add a fraction column
  quant_matrix = cbind(quant_meta,quant_sn) #combined the meta and final expression data
 
  ####this section merges all of the data together
  all_quant_files = rbind(all_quant_files, quant_matrix)
  
  message('finished ',i,' files') #output message for tracking progress
}
all_data = merge(psm_f3,all_quant_files, by=c('MS2ScanNumber','fraction'),all.x=TRUE) #bind the identification and quant data
#save the output data frame for later use
saveRDS(all_data, '/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant.rds')
```

Now we can do some analysis of the data, first looking at normalization. 

```{r}
id_table = readRDS('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant.rds') #grab the input data
id_sub1 = subset(id_table, id_table$totalSN > 100 & id_table$totalNA < 6) #filter by expression
quant_columns = c('tmt126_intensity','tmt127N_intensity','tmt127C_intensity','tmt128N_intensity','tmt128C_intensity',
                  'tmt129N_intensity','tmt129C_intensity','tmt130N_intensity','tmt130C_intensity') #make a vector of columns
id_sub2 <- setDT(id_sub1)[, lapply(.SD,sum,na.rm=TRUE),by=.(gene,sequence),.SDcols=quant_columns] #collapse identified peptides captured multiple times across different fractions
setDF(id_sub2) # convert back to dataframe

summary(id_sub2[,3:11]) #look at the data distributions to investigate normalization
#the data looks quite close
signal_sums = colSums(id_sub2[,3:11],na.rm=TRUE) #calculate total sum of signal to noise
#now loop over each of the columns taking the the proportion of the ion signal relative to the total multiplied by the median of all totals
for (l in 3:11){
		total_intensity = sum(id_sub2[,l],na.rm=TRUE)
		id_sub2[,l] = (id_sub2[,l] / total_intensity) * median(signal_sums)
}
id_sub2[,3:11][id_sub2[,3:11] == 0] = NA #it added some zeros, so lets get them back to NA
summary(id_sub2[,3:11]) #look at the summary again
#looks even better now
saveRDS(id_sub2,'/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant_normalized.rds')
```

Now we can look at gene expression in the normalized data. 

```{r}
id_sub2 = readRDS('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant_normalized.rds')
quant_columns = c('tmt126_intensity','tmt127N_intensity','tmt127C_intensity','tmt128N_intensity','tmt128C_intensity',
                  'tmt129N_intensity','tmt129C_intensity','tmt130N_intensity','tmt130C_intensity') #make a vector of columns
id_sub2[,3:11] = log2(id_sub2[,3:11]) #log transform the normalized expression values
id_sub3 <- setDT(id_sub2)[, lapply(.SD,median,na.rm=TRUE),by=.(gene),.SDcols=quant_columns] #compress to invididual genes using a sum
setDF(id_sub3) # convert back to dataframe

#make plots for the specific markers that Paulo showed (e.g. HXT3, SUC2, GAL10)
gal10 = melt(subset(id_sub3, grepl('GAL10',id_sub3$gene)))
gal10$sample = as.factor(c('gal','gal','gal','glu','glu','glu','raf','raf','raf'))
#prepare the plot
gal10_plot = ggplot(gal10, aes(sample, value)) +
  geom_boxplot() +
  geom_point(size=3,alpha=0.5,color='grey20') +
  scale_y_continuous(limits=c(4,12), breaks=seq(4,12,2)) +
  labs(x = "Sample", y = 'Median of Peptide Expression', title = 'GAL10 Expression')
gal10_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-sources-gal10-expression.pdf',gal10_plot) #save the plot
gal10_st = lm(value ~ C(sample,base=1), data = gal10) #calculate some statistics
summary(gal10_st) #view the linear model
confint(gal10_st) #calculate confidence intervals

suc2 = melt(subset(id_sub3, grepl('SUC2',id_sub3$gene)))
suc2$sample = as.factor(c('gal','gal','gal','glu','glu','glu','raf','raf','raf'))
#prepare the plot
suc2_plot = ggplot(suc2, aes(sample, value)) +
  geom_boxplot() +
  geom_point(size=3,alpha=0.5,color='grey20') +
  scale_y_continuous(limits=c(4,10), breaks=seq(4,10,2)) +
  labs(x = "Sample", y = 'Median of Peptide Expression', title = 'SUC2 Expression')
suc2_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-sources-suc2-expression.pdf',suc2_plot) #save the plot
suc2_st = lm(value ~ C(sample,base=3), data = suc2) #calculate some statistics
summary(suc2_st) #view the linear model
confint(suc2_st) #calculate confidence intervals

hxt3 = melt(subset(id_sub3, grepl('HXT3',id_sub3$gene)))
hxt3$sample = as.factor(c('gal','gal','gal','glu','glu','glu','raf','raf','raf'))
#prepare the plot
hxt3_plot = ggplot(hxt3, aes(sample, value)) +
  geom_boxplot() +
  geom_point(size=3,alpha=0.5,color='grey20') +
  scale_y_continuous(limits=c(3,12), breaks=seq(3,12,2)) +
  labs(x = "Sample", y = 'Median of Peptide Expression', title = 'HXT3 Expression')
hxt3_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-sources-hxt3-expression.pdf',hxt3_plot) #save the plot
hxt3_st = lm(value ~ C(sample,base=2), data = hxt3) #calculate some statistics
summary(hxt3_st) #view the linear model
confint(hxt3_st) #calculate confidence intervals
```

We can also look at global expression patterns using the published tables to get the correation values.

```{r, message = FALSE}
id_sub2 = readRDS('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant_normalized.rds') #grab the RawQuant data
#grab the paulo peptide expression table
paulo_peptides = read.table('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/expression-tables/SupplementalTable2.txt', sep='\t', header=TRUE) #had to remove some hash tags from this file
paulo_peptides$sequence = gsub('[^A-Z]','',sub(".*\\.([a-zA-Z]+)\\..*", "\\1", paulo_peptides$PeptideSequence)) #parse peptides
quant_columns = c('X126_sn','X127n_sn','X127c_sn','X128n_sn','X128c_sn','X129n_sn','X129c_sn','X130n_sn','X130c_sn')
paulo_sub <- setDT(paulo_peptides)[, lapply(.SD,sum,na.rm=TRUE),by=.(GeneSymbol,sequence),.SDcols=quant_columns] #collapse redundant hits, taking the sum across labels
setDF(paulo_sub) #convert back to dataframe
colnames(paulo_sub)[1] = 'gene'
paulo_scan = merge(paulo_sub,id_sub2,by=c('gene','sequence')) #merge with the RawQuant data

quant_matrix = as.matrix(cor(paulo_scan[,3:11],paulo_scan[,12:20],use='pairwise.complete.obs',method='pearson')) #make the correlation matrix
#set the range
mybreaks = seq(0.2,1,by=0.1)
#make the plot
#pdf('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-expression-correlation.pdf')
#make the correlation heatmap
heatmap.2(
		as.matrix(quant_matrix),
		col= rev(colorRampPalette(brewer.pal(6,"RdBu"))(length(mybreaks)-1)),
		symkey=FALSE,
		Rowv=FALSE,
		Colv=FALSE,
		dendrogram="none",
		breaks=mybreaks,
		cellnote = round(quant_matrix,2),
		notecol = 'black',
		#labRow = '',
		las=2,
		#ColSideColors=ColSideColors,
		## labels
		main='expression correlation',
		## color key
		key = TRUE,
		keysize = 1,
		density.info = "none",
		scale = "none",
		trace = "none",
		mar=c(8,8),
		cexRow=1.5,
		cexCol=0.75
)
#dev.off()
```

Look for the set of top 114 changers from the Paulo data in the RawQuant data.

```{r}
id_sub2 = readRDS('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-yeast_all-psms-id-quant_normalized.rds') #get the RawQuant data
id_sub2[,3:11] = log2(id_sub2[,3:11]) #log transform the expression data
quant_columns = c('tmt126_intensity','tmt127N_intensity','tmt127C_intensity','tmt128N_intensity','tmt128C_intensity',
                  'tmt129N_intensity','tmt129C_intensity','tmt130N_intensity','tmt130C_intensity') #make a vector of columns
id_sub3 <- setDT(id_sub2)[, lapply(.SD,median,na.rm=TRUE),by=.(gene),.SDcols=quant_columns] #collapse into genes
setDF(id_sub3) # convert back to dataframe
id_sub3$var = apply(id_sub3[,2:10], 1, function(x) var(x,na.rm=TRUE)) #calculate the variance across the samples
id_sub3 = id_sub3[order(-id_sub3$var),] #order by variance
paulo_114 = read.table('/projects/ptx_results/OtherDataSets/2016-April_Paulo_MolBioCell/expression-tables/SupplementalTable4.txt', sep='\t', header=TRUE) #grab the paulo 114 expression table
paulo_sub = paulo_114[,c(2,25)] #subset the paulo data
colnames(paulo_sub)[1] = 'gene'
#determine the intersection between the two gene lists through merging - want to look at correlation after
id_paulo = merge(paulo_sub,id_sub3,by='gene',all.x=TRUE) #determine the intersection between the two gene lists through merging

#count how many are missing from the RawQuant data
length(which(is.na(id_paulo$var))) #there seems to be 1 blank row - omit from calculation - 6 missing genes in total, none of which are actually found in the entire data set - search engine differences?

#what does the correlation look like between the two variations?
cor(id_paulo$CoV,id_paulo$var,use='pairwise.complete.obs',method='pearson') #only 0.7
#plot the values
var_correlation = ggplot(id_paulo, aes(log2(CoV), log2(var))) +
  geom_point(colour='grey20',size=5,alpha=0.5) +
  labs(x = "paulo", y = 'this study', title = 'Protein Variance Correlation')
var_correlation #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-top114_variance_correlation.pdf',var_correlation) #output plot
```

Perform some clustering of the data to see if we can recapitulate the observed values from the original paper.

```{r}
###make a pca of the carbon sources
eset = id_sub2[,3:11] #log transform the normalized expression values
pca = prcomp(t(eset[complete.cases(eset),])) #perform the pca analysis
pca_ggIN = as.data.frame(pca$x)[,1:2] #transform the data to an object compatible with ggplot
pca_ggIN$sample = c('gal','gal','gal','glu','glu','glu','raf','raf','raf')
summary(pca) #pull the summary to see the variation values
brew_red = brewer.pal(6,'YlOrRd')[5] #prepare some colors for lines
brew_points = brewer.pal(3,'Set2') #prepare colors for points
#prepare the plot
carbon_pca = ggplot(pca_ggIN, aes(PC1, PC2)) +
  geom_point(aes(colour=sample, shape=sample), size=5, alpha=0.5) +
  geom_hline(yintercept = 0, col=brew_red, linetype='dashed') +
  geom_vline(xintercept = 0, col=brew_red, linetype='dashed') +
  labs(x = "Component 1 (60% of variation)", y = 'Component 2 (34% of variation)', title = 'PCA of Carbon Sources') +
  scale_y_continuous(limits = c(-150,150), breaks = seq(-150,150,50)) +
  scale_x_continuous(limits = c(-200,250), breaks = seq(-200,250,50)) +
  scale_color_manual(values = brew_points)
  #theme(legend.position="none")
carbon_pca #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_MBC-2016/Routput/paulo-carbon-sources-protein-expression-pca.pdf',carbon_pca) #save the plot
```

Get the scan numbers.

```{r}
ms2_data = data.frame() #data storage object 
#iterate over the quantification files to get the numbers of MS2 scans
for (i in 1:length(ms2_quant_files)){
  ms2_in = read.table(ms2_quant_files[i],header=TRUE,sep='\t') #read in the file
  ms2_data = rbind(ms2_data,ms2_in) #bind the data
}
ms1_data = data.frame() #data storage object 
#iterate over the quantification files to get the numbers of MS1 scans
for (i in 1:length(ms1_quant_files)){
  ms1_in = read.table(ms1_quant_files[i],header=TRUE,sep='\t') #read in the file
  ms1_data = rbind(ms1_data,ms1_in) #bind the data
}
ms3_data = data.frame() #data storage object 
#iterate over the quantification files to get the numbers of MS3 scans
for (i in 1:length(quant_files)){
  ms3_in = read.table(quant_files[i],header=TRUE,sep='\t') #read in the file
  ms3_data = rbind(ms3_data,ms3_in) #bind the data
}

```

Done.

