---
title: "Analysis of Orbitrap Fusion Lumos MS3 Data from Paulo *et al.*"
author: "Christopher Hughes"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

###Description

This file contains the code for processing the search and quantification output files from the triple knockout publication from Paulo *et al.* (PMID: 27400695). **These data were used to make portions of Figure S-2 - S-5.**

  * MGF files were generated using RawQuant.
  * The data were searched using SearchCLI and PeptideShakerCLI.
  * Raw files were quantified with RawQuant and Proteome Discoverer in parallel.
  
###Analysis

First call the required libraries.

```{r, message=FALSE}
library(Biostrings)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(cowplot)
library(reshape)
```

Next we want to make an index based on our search database that we can use to add descriptive data to our search result.

```{r}
####
yeastDB = readAAStringSet('/projects/ptx_analysis/chughes/databases/uniprot_s288c-crap-oct2017-FWD.fasta')
#make an annotation index from yeastDB
yeastIndex = data.frame('meta' = names(yeastDB))
yeastIndex$accession = sub(".*?\\|(.*?)\\|.*", "\\1", yeastIndex$meta) #grab the accession from the header
yeastIndex$gene = sub(".*?GN=(.*?) .*", "\\1", yeastIndex$meta) #grab the gene from the header
yeastIndex$species = sub(".*?OS=(.*?) [GP].*", "\\1", yeastIndex$meta) #grab the species from the header
yeastIndex = yeastIndex[,2:4]
```

Now we can go ahead and grab the identification and quantification data.

```{r}
#find the files of interest
#psms files from SearchCLI and PeptideShaker
psm_files = list.files('/projects/ptx_results/OtherDataSets/2017-June_Paulo_TKO/reports-output',pattern='L.*ms3.*Default_PSM_Report.txt',full.names=TRUE)
#quant files from RawQuant algorithm
quant_files = list.files('/projects/ptx_results/OtherDataSets/2017-June_Paulo_TKO/quant-output',pattern='L.*ms3.*QuantData.txt',full.names=TRUE)
#quant files from Proteome Discoverer
pd_files = list.files('/projects/ptx_results/OtherDataSets/2017-June_Paulo_TKO/pd-quant-output',pattern='L.*ms3.*QuanSpectra.txt',full.names=TRUE)
```

Now we can process the peptide spectral match data to a more manageable data frame, and incorporate the quantification data from the two different sources.

```{r}
psm_all = data.frame() #create an empty data frame to hold the output data
#iterate over all the psm files
for (i in 1:length(psm_files)){
  
  ###this first section handles the identification data
  psm_in = read.table(psm_files[i],header=TRUE,sep='\t') #grab the input file
  psm_f1 = psm_in %>% 
    dplyr::select(Protein.s., Sequence, Spectrum.File, Spectrum.Title, m.z, Identification.Charge) %>% #select columns
    dplyr::mutate(Spectrum.Title, scan_num = sub(".*?spectrum (.*?)", "\\1", Spectrum.Title)) %>% #parse out the scan number
    dplyr::mutate(Identification.Charge, id_charge = gsub("[^0-9]", "", Identification.Charge)) %>% #parse out the charge
    dplyr::mutate(Spectrum.Title, fraction = sub("(.*?)_MGF\\.mgf", "\\1", Spectrum.File)) %>% #parse the fraction
    dplyr::select(Protein.s., Sequence, scan_num, m.z, id_charge, fraction) %>% #select new columns
    dplyr::mutate(Protein.s., first_accession = sapply(strsplit(as.character(Protein.s.),"\\,"), `[`, 1)) #parse the accession
  #adjust the column names
  colnames(psm_f1) = c('accession_groups','sequence','MS2ScanNumber','id_precursor_mz','id_charge','fraction','accession')
  psm_f2 = merge(psm_f1,yeastIndex,by='accession',all.x=TRUE) #add the descriptive data
  psm_f3 = dplyr::filter(psm_f2, !grepl('CONT', accession)) #remove contaminant hits
  
  
  ####this second section handles the RawQuant quantification data
  quant_in = read.table(quant_files[i],header=TRUE,sep='\t') #grab the input file
  quant_in[,11:90][quant_in[,11:90] == 0] = NA #transform any zero or missing values to NA
  quant_f1 = subset(quant_in, rowSums(is.na(quant_in[,c('tmt126_intensity',
                                  'tmt127N_intensity','tmt127C_intensity','tmt128N_intensity','tmt128C_intensity',
                                  'tmt129N_intensity','tmt129C_intensity','tmt130N_intensity','tmt130C_intensity',
                                  'tmt131_intensity')])) < 10) #calculate how many scans have zero reporters
  message('number of scans without reporter ion signals = ', nrow(quant_in) - nrow(quant_f1)) #output message
  quant_meta = quant_in[,c('PrecursorMass','ScanNumber','MS1ScanNumber','MS2ScanNumber','RetentionTime',
                           'PrecursorCharge','MS3IonInjectionTime','MS1Interference','tmt126_mass','tmt127N_mass',
                           'tmt127C_mass','tmt128N_mass','tmt128C_mass','tmt129N_mass','tmt129C_mass',
                           'tmt130N_mass','tmt130C_mass','tmt131_mass')] #extract the quantification meta data
  quant_exprs = quant_in[,c('tmt126_intensity','tmt127N_intensity','tmt127C_intensity','tmt128N_intensity',
                            'tmt128C_intensity','tmt129N_intensity','tmt129C_intensity','tmt130N_intensity',
                            'tmt130C_intensity','tmt131_intensity','tmt126_noise','tmt127N_noise',
                            'tmt127C_noise','tmt128N_noise','tmt128C_noise','tmt129N_noise','tmt129C_noise','tmt130N_noise',
                            'tmt130C_noise','tmt131_noise')] #extract the quantification data
  quant_sn = quant_exprs[,c(1:10)] / quant_exprs[,c(11:20)] #calculate the signal to noise
  quant_sn$totalIntensity = rowSums(quant_exprs[,1:10],na.rm=TRUE) #calculate the total sum intensity
  quant_sn$totalSN = rowSums(quant_sn[,1:10],na.rm=TRUE) #calculate the total sum signal to noise
  quant_sn$totalNoise = rowSums(quant_exprs[,11:20],na.rm=TRUE) #calculate the total sum noise
  quant_sn$totalSPS = rowSums(quant_in[,81:90],na.rm=TRUE) #calculate the total sum SPS ion intensity
  quant_sn$SPScount = rowSums(!is.na(quant_in[,71:80])) #calculate the number of SPS ions selected
  quant_sn$totalNA = rowSums(is.na(quant_sn[,1:10])) #grab the number of NA values across the TMT channels
  quant_meta$fraction = sub(".*?quant-output\\/(.*?)_Quant.*", "\\1", quant_files[i]) #add a fraction column
  quant_matrix = cbind(quant_meta,quant_sn) #combined the meta and final expression data
  
  
  
  ####this third section handles the Proteome Discoverer quantification data
  pd_in = read.table(pd_files[i],header=TRUE,sep='\t') #grab the input file
  colnames(pd_in) = c('Checked','BestPSMAmbiguity','Analyzer','ActivationType','MSOrder','NumberOfPSMs',
                      'pd_MS1Interference','pd_MS3IonInjectionTime','pd_PrecursorMass','mono','pd_PrecursorCharge',
                      'pd_RetentionTime','ScanNumber','fraction','pd_meanSN','pd_126','pd_127N','pd_127C','pd_128N',
                      'pd_128C','pd_129N','pd_129C','pd_130N','pd_130C','pd_131N','pd_131C') #reassign column names
  pd_in$fraction = paste(sub("(.*)\\.raw", "\\1", pd_in$fraction),'_ms3',sep='') #add a fraction column

  
  ####this section merges all of the data together
  psm_quant1 = merge(psm_f3,quant_matrix, by=c('MS2ScanNumber','fraction'),all.x=TRUE) #merge the identifications with RawQuant data
  psm_quant2 = merge(psm_quant1,pd_in,by=c('ScanNumber','fraction'),all.x=TRUE) #merge with Proteome Discoverer data
  psm_all = rbind(psm_all,psm_quant2) #bind all the data into a continuous single set
  
  message('finished ',i,' files') #output message for tracking progress
}
#save the output data frame for later use
saveRDS(psm_all, '/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo-lumos-ms3_all-psms-id-quant.rds')
```

Now we can plot some of the data to see how well the observations align with previously published results.

```{r}
psm_set = readRDS('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo-lumos-ms3_all-psms-id-quant.rds') #read in the processed data

#look at the known knockout genes
met6 = subset(psm_set, grepl('MET6',psm_set$gene))
pfk2 = subset(psm_set, grepl('PFK2',psm_set$gene))
ura2 = subset(psm_set, grepl('URA2',psm_set$gene))


#make a boxplot for met6 using the Proteome Discoverer values
met6pd_plot = ggplot(melt(met6[,c(56:64)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'Proteome Discoverer') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(met6[,c(56:58)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,250),breaks=seq(0,250,50))
met6pd_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_met6_lumosMS3-PD_boxplot.pdf',met6pd_plot) #save the plot


#make a boxplot for met6 using the RawQuant values
met6py_plot = ggplot(melt(met6[,c(27:35)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'RawQuant') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(met6[,c(27:29)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,250),breaks=seq(0,250,50))
met6py_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_met6_lumosMS3-PY_boxplot.pdf',met6py_plot) #save the plot


#make a boxplot for pfk2 using the Proteome Discoverer values
pfk2pd_plot = ggplot(melt(pfk2[,c(56:64)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'Proteome Discoverer') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(pfk2[,c(59:61)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,350),breaks=seq(0,350,50))
pfk2pd_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_pfk2_lumosMS3-PD_boxplot.pdf',pfk2pd_plot) #save the plot


#make a boxplot for pfk2 using the RawQuant values
pfk2py_plot = ggplot(melt(pfk2[,c(27:35)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'RawQuant') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(pfk2[,c(30:32)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,350),breaks=seq(0,350,50))
pfk2py_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_pfk2_lumosMS3-PY_boxplot.pdf',pfk2py_plot) #save the plot


#make a boxplot for ura2 using the Proteome Discoverer values
ura2pd_plot = ggplot(melt(ura2[,c(56:64)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'Proteome Discoverer') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(ura2[,c(62:64)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,100),breaks=seq(0,100,25))
ura2pd_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_ura2_lumosMS3-PD_boxplot.pdf',ura2pd_plot) #save the plot


#make a boxplot for ura2 using the RawQuant values
ura2py_plot = ggplot(melt(ura2[,c(27:35)]), aes(variable,value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha=0.5, size=0.8, colour='grey20') +
  labs(x = "TMT Channel", y = 'Signal-to-Noise', title = 'RawQuant') +
  theme(axis.text.x = element_text(hjust = 1, size = 9)) +
  geom_hline(yintercept=median(melt(ura2[,c(33:35)])$value,na.rm=TRUE),color=brewer.pal(9,'YlOrRd')[8],linetype='dashed',size=1) +
  scale_y_continuous(limits=c(0,100),breaks=seq(0,100,25))
ura2py_plot #call the plot
save_plot('/projects/ptx_analysis/chughes/projects-current/raw-quant/published-data/paulo_TKO/Routput/paulo_ura2_lumosMS3-PY_boxplot.pdf',ura2py_plot) #save the plot

#output the expression values for each of the plots
message(median(melt(met6[,c(56:58)])$value,na.rm=TRUE))
message(median(melt(met6[,c(27:29)])$value,na.rm=TRUE))
message(median(melt(pfk2[,c(59:61)])$value,na.rm=TRUE))
message(median(melt(pfk2[,c(30:32)])$value,na.rm=TRUE))
message(median(melt(ura2[,c(62:64)])$value,na.rm=TRUE))
message(median(melt(ura2[,c(33:35)])$value,na.rm=TRUE))
```
 
 Done.